Self update logic errors

Playing with using this in a new repo and I see some warts related to the self-update code.

Since I'm not familiar with the logic here I am just posting symptoms for now. I can provide details/reproduction wherever this isn't clear.

All issues were test & verified with the 3 new-enough CMake versions I have available - 3.29.1 & 3.31.5 on Linux


1. Partial hash seems to trip update logic even when it matches.
```
-- CICD: cached version (f0e0de27f3023cab22d6714e35b1c953a13d7391) is not requested version (f0e0de2); checking upstream
```

2. There are empty string errors from regex. Although due to environment restrictions I am not easily able to confirm this git command would work.
```
CMake Error at output/.cache/cpm/cicd-repo-infrastructure/c7c424571de3b90318c6be58ed4cee9efd788b73/CMakeLists.txt:56 (string):
  string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command.
Call Stack (most recent call first):
  output/.cache/cpm/cicd-repo-infrastructure/c7c424571de3b90318c6be58ed4cee9efd788b73/CMakeLists.txt:94 (cicd_check_self_update)
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -36,8 +36,17 @@ function(cicd_check_self_update)
         COMMAND ${GIT_PROGRAM} rev-parse HEAD
         WORKING_DIRECTORY ${my_path}
         OUTPUT_VARIABLE head_hash
+        RESULT_VARIABLE hash_ok
         OUTPUT_STRIP_TRAILING_WHITESPACE)
-    if(head_hash STREQUAL ${CPM_ARGS_GIT_TAG})
+    if(NOT hash_ok EQUAL 0)
+        message(
+            FATAL_ERROR
+                "CICD: error discovering the current hash (`${GIT_PROGRAM} rev-parse HEAD`) in ${my_path}"
+        )
+        return()
+    endif()
+    string(FIND ${head_hash} ${CPM_ARGS_GIT_TAG} hash_pos)
+    if(hash_pos EQUAL 0)
         return()
     endif()
 
@@ -48,12 +57,26 @@ function(cicd_check_self_update)
 
     # what's the hash for what you asked for?
     execute_process(
-        COMMAND ${GIT_PROGRAM} ls-remote -h ${CPM_ARGS_GIT_REPOSITORY}
+        COMMAND ${GIT_PROGRAM} ls-remote ${CPM_ARGS_GIT_REPOSITORY}
                 ${CPM_ARGS_GIT_TAG}
+        COMMAND cut -f1
         WORKING_DIRECTORY ${my_path}
         OUTPUT_VARIABLE remote_hash
+        RESULT_VARIABLE hash_ok
         OUTPUT_STRIP_TRAILING_WHITESPACE)
-    string(REGEX REPLACE "\t.*" "" remote_hash ${remote_hash})
+    if(NOT hash_ok EQUAL 0)
+        message(
+            FATAL_ERROR
+                "CICD: error discovering the remote hash (`${GIT_PROGRAM} ls-remote ${CPM_ARGS_GIT_REPOSITORY} ${CPM_ARGS_GIT_TAG}`) in ${my_path}"
+        )
+        return()
+    else()
+        # success and empty output from git ls-remote probably means we asked
+        # for a hash, not a branch/tag
+        if(remote_hash STREQUAL "")
+            set(remote_hash ${CPM_ARGS_GIT_TAG})
+        endif()
+    endif()
     message(STATUS "CICD: requested version hash is ${remote_hash}")
 
     # do we already fulfil that hash?
@@ -81,8 +104,8 @@ function(cicd_check_self_update)
             RESULT_VARIABLE reset_ok)
         if(NOT reset_ok EQUAL 0)
             message(
-                FATAL
-                "CICD: Couldn't update to ${remote_hash} - check the repository at ${my_path}."
+                FATAL_ERROR
+                    "CICD: Couldn't update to ${remote_hash} - check the repository at ${my_path}."
             )
         endif()
         message(STATUS "CICD: cached version is now at ${remote_hash}")

--- a/cmake/dependencies.cmake
+++ b/cmake/dependencies.cmake
@@ -229,8 +229,8 @@ function(update_versioned_package)
             RESULT_VARIABLE reset_ok)
         if(NOT reset_ok EQUAL 0)
             message(
-                FATAL
-                "Couldn't update ${ARGS_NAME} to ${ARGS_GIT_TAG} - check the repository at ${pkg_dir}."
+                FATAL_ERROR
+                    "Couldn't update ${ARGS_NAME} to ${ARGS_GIT_TAG} - check the repository at ${pkg_dir}."
             )
         endif()
     else()

